src/auth/src/retry.rs

```diff
@@ -103,31 +152,94 @@ where
+
+            // Create the intermediate error message to send out.
+            let cred_error = Self::map_retry_error(owned_error, Some(delay));
+            // Send the intermediate error.
+            tx_for_retry.send(Some(Err(cred_error))).ok();
+        };
+
+        let final_result = retry_loop_with_callback(
             fetch_token,
             sleep,
             true, // token fetching is idempotent
             self.retry_throttler.clone(),
-            retry_policy,
+            self.retry_policy.clone(),
             self.backoff_policy.clone(),
+            on_retry,
         )
-        .await
-        .map_err(Self::map_retry_error)
+        .await;
+
+        // Send the final result, whether success or a terminal error.
+        match final_result {
+            Ok(token) => {
+                self.token_watch_tx.send(Some(Ok(token))).ok();
+            }
+            Err(e) => {
+                // The final error, this can only be non-transient error.
+                let final_error = Self::map_retry_error(e, None);
+                self.token_watch_tx.send(Some(Err(final_error))).ok();
+            }
+        }
     }
 
-    fn map_retry_error(e: gax::error::Error) -> CredentialsError {
-        if !e.is_authentication() {
-            return CredentialsError::from_source(false, e);
+    fn map_retry_error(e: gax::error::Error, last_delay: Option<Duration>) -> CredentialsError {
+        let is_transient = last_delay.is_some();
+        let mut cred_error = CredentialsError::new(is_transient, e.to_string(), e);
+
+        if let Some(delay) = last_delay {
+            cred_error = cred_error.with_retry_in(delay);
         }
+        cred_error
+    }
+}
 
-        let msg = match e
-            .source()
-            .and_then(|s| s.downcast_ref::<CredentialsError>())
-        {
-            Some(cred_error) if cred_error.is_transient() => constants::RETRY_EXHAUSTED_ERROR,
-            _ => constants::TOKEN_FETCH_FAILED_ERROR,
-        };
-        CredentialsError::new(false, msg, e)
+#[async_trait::async_trait]
+impl TokenProvider for TokenProviderWithRetry {
+    async fn token(&self) -> Result<Token> {
+        let mut rx = self.token_watch.clone();
+
+        // The caller is requesting a new token, so we trigger a refresh.
+        // A send can fail if the channel is full or if the receiver has been
+        // dropped. If the channel is full, it means a refresh is already in
+        // progress. If the receiver has been dropped, it means the background
+        // task has terminated.
+        if self.refresh_trigger.try_send(()).is_err() {
+            if self.refresh_trigger.is_closed() {
+                return Err(CredentialsError::from_msg(
+                    false,
+                    "token provider background task has been terminated",
+                ));
+            }
+        }
+
+        // Wait for the next value to be published.
+        if rx.changed().await.is_err() {
```

Comment:
what if a token was pushed to channel between the time `refresh_trigger` send failed (i.e. there was an already existing retry loop going on) and `rx.changed` is called? The thread will wait until next token is available.

IIRC, the correct way is to first do rx.borrow so that rx knows what your code has already seen. Maybe just do borrow and see if there is new token. You could use entity tag system to know if it is a new token or existing token.